File: rails-server/udp_server.rb
require 'logger'
require 'socket'
require 'json'
require 'net/http'
require 'uri'
require 'websocket-client-simple'

udp_logger = Logger.new('log/udp_server.log')
udp_logger.level = Logger::DEBUG

# Configuraciones del servidor
API_URL = 'http://localhost:8080'.freeze
REGISTER_ENDPOINT = '/auth'.freeze
LOGIN_ENDPOINT = '/auth/sign_in'.freeze
WEBSOCKET_ENDPOINT = '/cable'.freeze
UDP_UID = 'udp@server.com'.freeze
UDP_PASSWORD = 123_456_789
# Mapea las acciones recibidas a los mensajes UDP para la Wemos D1
ACTION_TO_UDP = {
  turn_on:  { word: 'C_connect' },
  turn_off: { word: 'C_disconnect' }
}.freeze
# Configuración del servidor UDP para Wemos
WEMOS_IP = '192.168.1.70'.freeze # IP de la Wemos D1
WEMOS_PORT = 8080 # Puerto UDP de la Wemos D1

# Registra al usuario
def register_user
  uri = URI.parse("#{API_URL}#{REGISTER_ENDPOINT}")
  http = Net::HTTP.new(uri.host, uri.port)
  request = Net::HTTP::Post.new(uri.path, { 'Content-Type' => 'application/json' })
  request.body = { user: { email: UDP_UID, password: UDP_PASSWORD, password_confirmation: UDP_PASSWORD } }.to_json

  response = http.request(request)
  JSON.parse(response.body)
end

# Realiza login y obtiene los tokens
def login_user
  uri = URI.parse("#{API_URL}#{LOGIN_ENDPOINT}")
  http = Net::HTTP.new(uri.host, uri.port)
  request = Net::HTTP::Post.new(uri.path, { 'Content-Type' => 'application/json' })
  request.body = { email: UDP_UID, password: UDP_PASSWORD }.to_json

  response = http.request(request)
  headers = response.to_hash

  # Obtiene uid, client y access-token del header de la respuesta
  {
    uid:          headers['uid'].first,
    client:       headers['client'].first,
    access_token: headers['access-token'].first
  }
end

# Conecta al WebSocket con autenticación
def connect_websocket(auth_tokens)
  websocket_url = 'ws://localhost:8080/cable'
  access_token = auth_tokens[:access_token]
  client = auth_tokens[:client]
  uid = auth_tokens[:uid]
  websocket_params = "access-token=#{access_token}&client=#{client}&uid=#{uid}"
  ws = WebSocket::Client::Simple.connect "#{websocket_url}?#{websocket_params}"

  ws.on :message do |msg|
    udp_logger.info "Mensaje recibido en WebSocket: #{msg.data}"
    begin
      data = JSON.parse(msg.data)
      action = data['action']

      if ACTION_TO_UDP.key?(action)
        udp_message = ACTION_TO_UDP[action].to_json
        send_udp_message_to_wemos(udp_message)
        udp_logger.info "Mensaje UDP enviado a Wemos: #{udp_message}"
      else
        udp_logger.info "Acción no reconocida en WebSocket: #{action}"
      end
    rescue JSON::ParserError
      udp_logger.info "Mensaje de WebSocket no válido: #{msg.data}"
    end
  end

  ws.on :open do
    udp_logger.info 'Conexión WebSocket abierta'
  end

  ws.on :close do
    udp_logger.info 'Conexión WebSocket cerrada'
  end

  ws.on :error do |e|
    udp_logger.info "Error en WebSocket: #{e.message}"
  end

  ws
end

# Función para enviar mensaje UDP a la Wemos
def send_udp_message_to_wemos(message)
  udp_socket = UDPSocket.new
  udp_socket.send(message, 0, WEMOS_IP, WEMOS_PORT)
  udp_socket.close
end

# Inicia el registro y login para obtener el token
udp_logger.info 'Registrando y autenticando al usuario...'
register_user # Registra el usuario
auth_tokens = login_user # Realiza login y obtiene las claves

# Conexión al WebSocket
ws = connect_websocket(auth_tokens)

# Configura el socket UDP para escuchar en el puerto 12345
udp_socket = UDPSocket.new
udp_socket.bind('0.0.0.0', 12_345)

# Límite de bytes para recibir
MAX_BYTES = 65_507

udp_logger.info 'UDP Server listening on port 12345...'

# Loop infinito para recibir mensajes UDP y enviarlos por WebSocket
loop do
  # Recibir mensaje UDP
  message, sender = udp_socket.recvfrom(MAX_BYTES)
  sender_ip = sender[3]
  sender_port = sender[1]

  udp_logger.info "Mensaje recibido: '#{message.strip}' de #{sender_ip}:#{sender_port}"

  # Formatear el mensaje como JSON para enviar a través de WebSocket
  payload = {
    message: message.strip,
    ip:      sender_ip,
    port:    sender_port
  }.to_json

  # Enviar el mensaje al WebSocket
  ws.send(payload)
end

File: udp-server/udp_server.py
import json
import socket
from datetime import datetime
import requests
from websocket import create_connection
import os

# Constantes para el webserver
API_URL = os.getenv("WEBSERVER_URL", "http://webserver")
WEBSOCKET_URL = os.getenv("WEBSOCKET_URL", "ws://webserver")
WEBSOCKET_PORT = int(os.getenv("WEBSOCKET_PORT", 8080))
REGISTER_ENDPOINT = "/auth"
LOGIN_ENDPOINT = "/auth/sign_in"
WEBSOCKET_URI = "/cable"

# Constantes para la Wemos
WEMOS_IP = os.getenv("WEMOS_IP", "192.168.1.65")
WEMOS_PORT = int(os.getenv("WEMOS_PORT", 44444))

# Constantes para UDP
UDP_PORT = 12345
UDP_UID = "udp@server.com"
UDP_PASSWORD = "123456789"


def log(message):
  print(f"{datetime.now().isoformat()} - {message}")

# Registra al usuario


def register_user():
  log("Registrando usuario...")
  url = f"{API_URL}{REGISTER_ENDPOINT}"
  payload = {
      "user": {
          "email": UDP_UID,
          "password": UDP_PASSWORD,
          "password_confirmation": UDP_PASSWORD,
      }
  }
  response = requests.post(url, json=payload)
  if response.status_code == 200:
    log("Registro exitoso.")
  else:
    log(f"Error al registrar usuario: {response.text}")

# Realiza login y obtiene los tokens


def login_user():
  log("Autenticando usuario...")
  url = f"{API_URL}{LOGIN_ENDPOINT}"
  payload = {"email": UDP_UID, "password": UDP_PASSWORD}
  response = requests.post(url, json=payload)
  if response.status_code == 200:
    log("Autenticación exitosa.")
    headers = response.headers
    return {
        "access_token": headers.get("access-token"),
        "client": headers.get("client"),
        "uid": headers.get("uid"),
    }
  else:
    log(f"Error en la autenticación: {response.text}")
    return None

# Conecta al WebSocket con autenticación


def connect_websocket(auth_tokens):
  log("Conectando al WebSocket...")
  websocket_url = f"{WEBSOCKET_URL}{WEBSOCKET_URI}:{WEBSOCKET_PORT}"
  query_params = f"access-token={auth_tokens['access_token']}&client={auth_tokens['client']}&uid={auth_tokens['uid']}"
  print(f"Conectando a WebSocket en: {websocket_url}?{query_params}")
  ws = create_connection(f"{websocket_url}?{query_params}")

  log("Conexión WebSocket establecida.")
  return ws

# Envía un mensaje UDP a la Wemos


def send_udp_message_to_wemos(message):

  print("Mensaje: " + message)
  datos = bytes.fromhex(message[2:])  # Skip '0x' prefix
        
  # Crear socket UDP
  sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
  
  # Enviar datos
  sock.sendto(datos, (WEMOS_IP, WEMOS_PORT))

  log(f"Mensaje UDP enviado a Wemos: {datos}")


# Inicia el servidor UDP y reenvía mensajes al WebSocket


def start_udp_server(auth_tokens):
  ws = connect_websocket(auth_tokens)

  udp_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
  udp_socket.bind(("0.0.0.0", UDP_PORT))

  log(f"Servidor UDP escuchando en el puerto {UDP_PORT}...")

  while True:
    message, sender = udp_socket.recvfrom(65507)
    sender_ip, sender_port = sender
    message = message.decode().strip()
    log(f"Mensaje recibido: '{message}' de {sender_ip}:{sender_port}")

    # Envía directamente el mensaje a la Wemos
    send_udp_message_to_wemos(message)


if __name__ == "__main__":
  # Registra y autentica al usuario
  register_user()
  auth_tokens = login_user()

  if auth_tokens:
    # Inicia el servidor UDP
    start_udp_server(auth_tokens)
  else:
    log("No se pudo autenticar al usuario. El servidor no se iniciará.")